pkg/binary/envoy/debug/pii.go

```diff
@@ -0,0 +1,72 @@
+package debug
+
+import (
+	"fmt"
+	"bitbucket.org/creachadair/shell"
+	"crypto/sha256"
+)
+
+// filter the array of logs using formatStr and validFieldNames, hash the resulting array of logs and return the final array
+func process(logsArr []string, formatStr string, validFieldNames map[string]bool) (processedLogs []string) {
+	defer func() {
+		if r := recover(); r != nil {
+			fmt.Println("process func panicking, recovering")
+		}
+	} ()
+
+	fieldNames, ok := shell.Split(formatStr)
+
+	if !ok {
+		panic(fmt.Sprint("error in splitting format string: %s", formatStr))
+	}
+
+	testForValidLog :=  func(log string) bool {
+		fields, ok := shell.Split(log)
+		if !ok {
+			panic(fmt.Sprint("error in splitting log: %s", log))
+		}
+		// TODO: consider using regex to check fields instead of merely verifiying the number of fields
+		return len(fields) == len(fieldNames)
+	}
+	validLogs := filter(logsArr, testForValidLog)
+
+	// filter and hash each log in the array of valid logs with relevant fields
+	for i,_ := range(validLogs) {
+		fields, ok := shell.Split(validLogs[i])
+		if !ok {
+			panic(fmt.Sprint("error in splitting a valid log: %s", validLogs[i]))
+		}
+
+		// filter fields by validFieldNames
+		filteredFields := []string{}
+		for j,_ := range(fields) {
+			if (validFieldNames[fieldNames[j]]) {
+				filteredFields = append(filteredFields,fields[j])
+			}
+		}
```

Comment:
We don't want to remove the PII fields then hash the entire log line, we only want to hash the PII fields. So this loop should look something like:

```go
for j, field := range fieldNames {
	if !containsPII(field) {
		// in golang, we prefer to keep the function as un-nested as possible. Write the other version and compare
		continue
	}
	// safe to index into fields with j because we checked fieldNames and fields have the same length above
	fields[j] = hash(fields[j])
}
processed = append(processed, strings.Join(fields))
```

Reply:
Also note that I flip this around and drive it by the format string, rather than driving it by the log line. This makes more sense, as you're matching up the log line to the format, not the format to the log.

Reply:
One final note: stylistically I might actually not use `continue` here and tuck the logic into the `if`, since it winds up only being 1 line:
```go
for j, field := range fieldNames {
	if containsPII(field) {
		// safe to index into fields with j because we checked fieldNames and fields have the same length above
		fields[j] = hash(fields[j])
	}
}
processed = append(processed, strings.Join(fields))
```

Reply:
Can I clarify if the final logs should have all the none PII fields removed?

Reply:
We want to call some function (our `hash` function for now) on each piece of the PII data and put the functions result into the log line instead of the original PII data. In the example I showed, we split the string into parts and replace the parts with PII with the result of calling `hash`.

In other words, given
```go
// Given: SOURCE_IP is PII, RESPONSE_CODE is not
fmtString := `%RESPONSE_CODE% %SOURCE_IP%`
logLines := []string{
	"200 127.0.0.1",
	"500 10.0.1.1",
	"404 192.168.17.0",
}

out := process(fmtString, logLines)
// should return the log lines with PII part scrambles
reflect.DeepEqual(out, []string{
	"200 AAB5=",
	"500 AS0978B",
	"404 BKL9P84",
})
```
(In a future iteration we'll extract out the call to `hash` so we can substitute in `hash` for other functions - for example replace IP addresses with `IP_ADDRESS` rather than random data - but don't worry about that yet.)

Reply:
I see, thanks for the detailed explanation